### PR TITLE
Exit codes

### PR DESCRIPTION
**Pull request by eduardobeloni**
We wanted to add proper return codes to the xfreerdp client. So there's a set of return codes in libfreerdp to be matched to those in the ui client, the last thread will provide the return code.

Please make your comments and suggestions on these commits.

The next step is to work on the man page for the return values.

### CODE REVIEW

**Comment by kiilerix** [comment]
It could be nice to have more consistency in the naming - is it disc_reason or disconnect_reason or (str)error or code or reason.

ERRINFO_MAX_BUFLEN ... isn't that more like a minimum or recommended length? It doesn't look like a max.

I wonder if it wouldn't be simpler to make disc_reason a global variable that all threads writes to. That might be less fragile than having the global g_last_thread.

match_disconnect_reason could have a more descriptive name such as exit_code_from_disc_reason. It would be a nice hint about the valid value range if it returned uint8. I think the function just should do the translation from reason code to exit code - it shouldn't print.

All the xfreerdp code should also be used in wfreerdp. The right solution to that is however neither to put it in a library nor cut'n'paste it but to merge these two tools (and dfbfreerdp).

**Comment by otavio** [comment]
About "disc_reason" naming I agree. Personally I dislike disc_reason and disconnect_reason seems more appropriate.

ERRINFO_BUFFER_SIZE seems a good name IMO.

The problem with having it as a global is about race conditions. Personally I think it ought to be done by thread. The X11 client can have it controlled in whatever why but the lib shouldn't use a global for it IMO.

exit_code_from_disc_reason (or disconnect_reason) or even disconnect_reason_to_exit_code seems better indeed.

About code duplication between wfreerdp and dfbfreerdp it seems will be better when libfreerdpgdi comes to a mature state. So I think we shouldn't botter too much on it now.

**Comment by kiilerix** [comment]
On 03/17/2011 07:29 PM, otavio wrote:

> The problem with having it as a global is about race conditions. Personally I think it ought to be done by thread. The X11 client can have it controlled in whatever why but the lib shouldn't use a global for it IMO.

Yes, the library should be purely functional - I only meant a global 
variable in the client code instead of a global threadid. And I guess 
the client is racy anyway if two sessions disconnects exactly at the 
same time.

**Comment by eduardobeloni** [comment]
I just updated the branch with the suggested modifications:

Consistent naming system;
Disconnect_reason is global istead of the last thread ID;
A few bug fixes.

File: libfreerdp/errinfo.c [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+   Error information
+
+   Copyright 2010 O.S. Systems Software Ltda.
+   Copyright 2010 Eduardo Fiss Beloni <beloni@ossystems.com.br>
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licen1ses/LICENSE-2.0
```
Comment (otavio) [inline libfreerdp/errinfo.c]: Typo on the URL

**Comment by eduardobeloni** [comment]
Well then,
The last commit treats exits code as if there's only one thread of execution.

After all, the exit code is neither reliable nor usable if more than one connections are created, like kiilerix said.

**Comment by otavio** [comment]
This seems OK for me so if there're no objections I'd like to get it merged so we can move on.

**Pull request closed without merging**